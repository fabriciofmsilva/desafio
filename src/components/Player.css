.Player {
  min-width: 240px;
  margin-right: 2rem;
  margin-left: 2rem;
  padding: 2rem;
  background-color: #ecf0f1;
}

.Player.dead {
  opacity: .5;
  transition: opacity .5s ease-in-out;
}

.Player-title {
  margin-top: 0;
  margin-bottom: .5rem;
  font-size: 2rem;
}

.Player-life {
  width: 100%;
  padding: .25rem 0;
  margin-bottom: .5rem;
  font-size: 1.2rem;
  color: #ecf0f1;
  background-color: #27ae60;
  transition: width .2s ease;
}

.Player-life.warning {
  background-color: #e67e22;
}

.Player-life.danger {
  background-color: #e74c3c;
}

.Player-sprite-1 {
  width: 64px;
  height: 64px;
  margin-left: auto;
  background-image: url("../img/player1.png");
}

.Player-sprite-2 {
  width: 64px;
  height: 64px;
  background-image: url("../img/player2.png");
}

.walk-right {
  animation: walk-right .8s steps(8) infinite;
}

.atack-right {
  animation: atack-right .8s steps(5) infinite;

  background-position: 0 -960px;
}

.walk-left {
  animation: walk-left .8s steps(8) infinite;
}

.atack-left {
  animation: atack-left .8s steps(5) infinite;
}

.dead {
  background-position: -330px -1286px;
  animation: dead .5s steps(5) 1;
}

@keyframes walk-left {
  from {
    background-position: 0 -580px;
  }

  to {
    background-position: -512px -580px;
  }
}

@keyframes atack-left {
  from {
    background-position: 0 -836px;
  }

  to {
    background-position: -320px -836px;
  }
}

@keyframes walk-right {
  from {
    background-position: 0 -706px;
  }

  to {
    background-position: -512px -706px;
  }
}

@keyframes atack-right {
  from {
    background-position: 0 -964px;
  }

  to {
    background-position: -320px -964px;
  }
}

@keyframes dead {
  from {
    background-position: 0px -1286px;
  }

  to {
    background-position: -330px -1286px;
  }
}
