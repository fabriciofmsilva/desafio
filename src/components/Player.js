import React from 'react';
import { connect } from 'react-redux';

import { atack } from '../actions/game';

import './Player.css';

export class Player extends React.Component {
  state = {
    state: 'walk'
  };

  getPlayerState = () => {
    if (this.props.player.life === 0) {
      return 'dead';
    }

    return this.props.player.id === 1 ? this.state.state + '-right' : this.state.state + '-left';
  }

  getLifeColor = (life) => {
    return life < 60 ? (life < 40 ? "danger" : "warning") : "";
  };

  atack = () => {
    this.setState((prevState, props) => {
      return {
        state: 'atack'
      };
    });

    const atackDelay = setTimeout(() => {
      this.props.atack(this.props.player.id);
      this.setState((prevState, props) => {
        clearTimeout(atackDelay);
        return {
          state: 'walk'
        };
      });
    }, 800);
  };

  render() {
    return (
      <div className={`Player ${this.props.player.life === 0 ? 'dead' : ''}`}>
        <h2 className="Player-title">
          {this.props.player.name}
        </h2>
        <div className={`Player-life ${this.getLifeColor(this.props.player.life)}`}
          style={{ width: `${this.props.player.life}%` }}
        >
          {this.props.player.life}%
        </div>
        <div className={`Player-sprite-${this.props.player.id} ${this.getPlayerState()}`}></div>
        <button
          className="button button-danger"
          disabled={this.props.gameover || this.state.state === 'atack'}
          onClick={this.atack}
        >
          Atacar
        </button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  player: state.game.players[props.player.id - 1],
  gameover: state.game.over
});

const mapDispatchToProps = (dispatch, props) => ({
  atack: (playerId) => {dispatch(atack(playerId))}
});

export default connect(mapStateToProps, mapDispatchToProps)(Player);
