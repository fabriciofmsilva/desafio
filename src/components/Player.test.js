import React from 'react';
import { shallow } from 'enzyme';

import { Player } from './Player';
import gameDefaultState from '../test/fixtures/game';

let wrapper;
let atack;

beforeEach(() => {
  atack = jest.fn();
  wrapper = shallow(<Player player={gameDefaultState.players[0]} />);
});

test('should render Player correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

// test('should handle onClick', () => {
//   wrapper.find('button').simulate('click');
//   // console.log(wrapper.props());
//   expect(atack).toHaveBeenCalled();
// });
