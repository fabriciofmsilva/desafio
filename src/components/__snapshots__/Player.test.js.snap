// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Player correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Player
    player={
        Object {
            "id": 1,
            "life": 100,
            "name": "Jogador 1",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="Player-life "
          style={
                    Object {
                              "width": "100%",
                            }
          }
>
          100
          %
</div>,
        <h2
          className="Player-title"
>
          Jogador 1
</h2>,
        <div
          className="Player-sprite-1 walk-right"
/>,
        <button
          className="button button-danger"
          disabled={false}
          onClick={[Function]}
>
          Atacar
</button>,
      ],
      "className": "Player ",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            100,
            "%",
          ],
          "className": "Player-life ",
          "style": Object {
            "width": "100%",
          },
        },
        "ref": null,
        "rendered": Array [
          100,
          "%",
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Jogador 1",
          "className": "Player-title",
        },
        "ref": null,
        "rendered": "Jogador 1",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "className": "Player-sprite-1 walk-right",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Atacar",
          "className": "button button-danger",
          "disabled": false,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Atacar",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="Player-life "
            style={
                        Object {
                                    "width": "100%",
                                  }
            }
>
            100
            %
</div>,
          <h2
            className="Player-title"
>
            Jogador 1
</h2>,
          <div
            className="Player-sprite-1 walk-right"
/>,
          <button
            className="button button-danger"
            disabled={false}
            onClick={[Function]}
>
            Atacar
</button>,
        ],
        "className": "Player ",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              100,
              "%",
            ],
            "className": "Player-life ",
            "style": Object {
              "width": "100%",
            },
          },
          "ref": null,
          "rendered": Array [
            100,
            "%",
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Jogador 1",
            "className": "Player-title",
          },
          "ref": null,
          "rendered": "Jogador 1",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "className": "Player-sprite-1 walk-right",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Atacar",
            "className": "button button-danger",
            "disabled": false,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Atacar",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
